### DISCUSSION TITLE
Why can't Mobx track dependencies in runInAction?

### DISCUSSION THREAD

**Discussion by janryWang** [mobxjs/mobx#2815] [Q&A] [answered]
Like this.

```ts
import { observable, autorun, runInAction } from "mobx";
const obs = observable({
  aa: 123
});

autorun(() => {
  runInAction(() => {
    if (obs.aa > 300) {
      obs.bb = 100;
    }
  });
});

obs.aa = 500;
console.log(obs.bb);
```

I did not find in the documentation that runInAction does not track dependencies，Is this a bug?

**Comment by kubk** [comment]
Why are you using autorun here? Your code would be much easier to follow without it:

```js
const obs = observable({
  aa: 123
});

function setAA(value) {
  obs.aa = value;
  if (obs.aa > 300) {
     obs.bb = 100;
  }
}

```

**Comment by mweststrate** [comment, marked as answer]
That is by design, see: https://mobx.js.org/the-gist-of-mobx.html#principles.
Your pattern isn't forbidden (but you have to put the condition outside the
action), but it is a bit of an anti pattern as it can easily create hard to
debug code-cycles. For example because bb would only be updated after the
current action (when the reaction runs), and not immediately after updating
aa, so you can observe an inconsistent state. We recommend to either
express it as a computed if that makes sense, or otherwise use a dedicated
function to update aa, and update bb as well with it.

**Comment by iChenLei** [comment]
Same as #2816 

Maybe you want to find the implemention of mobx souce code.

https://github.com/mobxjs/mobx/blob/851be05733926cba927864647c643497b8ac1142/packages/mobx/src/core/action.ts#L108-L110
